sycl/source/detail/queue_impl.cpp

```diff
@@ -234,21 +242,42 @@ void queue_impl::wait(const detail::code_location &CodeLoc) {
   TelemetryEvent = instrumentationProlog(CodeLoc, Name, StreamID, IId);
 #endif
 
-  std::vector<std::weak_ptr<event_impl>> Events;
-  std::vector<event> USMEvents;
+  std::vector<std::weak_ptr<event_impl>> WeakEvents;
+  std::vector<event> SharedEvents;
   {
-    std::lock_guard<std::mutex> Lock(MMutex);
-    Events.swap(MEventsWeak);
-    USMEvents.swap(MEventsShared);
+    std::lock_guard<mutex_class> Lock(MMutex);
+    WeakEvents.swap(MEventsWeak);
+    SharedEvents.swap(MEventsShared);
+  }
+  // If the queue is either a host one or does not support OOO (and we use
+  // multiple in-order queues as a result of that), wait for each event
+  // directly. Otherwise, only wait for unenqueued or host task events, starting
+  // from the latest submitted task in order to minimize total amount of calls,
```

Comment:
> starting from latest submitted task in order to minimize total amount of calls

Is the following case handled?

```
D ->C->B->A
E->A
Chronological order of submission: A, B, C, D, E.
notion: Dependant -> Dependency
```

Reply:
Yes, it is. I'm not actually sure whether waiting for E guarantees that D has already been enqueued by the time we start checking it (it seems that way but I might be getting lucky), but in the worst case scenario we'll wait for E and D here.